src/uri.c

```diff
@@ -57,15 +58,13 @@ static void decode(char *s)
 			/* abort on invalid format and chars */
 			if ((strlen(rp) < 3) ||
 			    !strchr(HEX_CHARS, rp[1]) ||
-			    !strchr(HEX_CHARS, rp[2])) {
-				*wp = '\0';
-				return;
-			}
+			    !strchr(HEX_CHARS, rp[2]))
+				goto out;
 
 			rp++;				/* skip % */
-			endptr = rp + 2;		/* 2 chars only */
+			memcpy(hex, rp, 2);		/* 2 chars only */
 
-			tmp = strtoul(rp, &endptr, 16);	/* convert */
+			tmp = strtoul(hex, NULL, 16);	/* convert */
```

Comment:
Instead of manually checking for HEX_CHARS above you could also let `strtoul()` do this for you by using the `endptr` parameter. You could check if after the call to `strtoul()`, if `*endptr == '\0'`.  `strtoul()` sets `endptr` to the first invalid character. So if both characters where converted, then `endptr` points to the 3rd byte of `hex` which is the ending zero byte. If the 2 hex chars contained any non-hex characters, then it would point to the first or the second byte of `hex`, and thus `*endptr` would not be `'\0'`.

Reply:
ok